recipes/dnas/zomes/storage/src/dnaversions.rs

```diff
@@ -0,0 +1,200 @@
+	chunk_addresses: input.chunk_addresses,
+	changelog: match input.changelog {
+	    None => String::from(""),
+	    Some(x) => x,
+	},
+	contributors: match input.contributors {
+	    None => vec![],
+	    Some(c) => c,
+	},
+	published_at: match input.published_at {
+	    None => {
+		sys_time()?.as_millis() as u64
+	    },
+	    Some(t) => t,
+	},
+	last_updated: match input.published_at {
+	    None => {
+		sys_time()?.as_millis() as u64
+	    },
+	    Some(t) => t,
+	},
+    };
+
+    let header_hash = create_entry(&version)?;
+    debug!("Created new DNA Version via header ({})", header_hash );
+
+    let entry_hash = hash_entry(&version)?;
+
+    debug!("Linking DNA ({}) to manifest: {}", input.for_dna, entry_hash );
+    create_link(
+	input.for_dna,
+	entry_hash.clone(),
+	LinkTag::new(*TAG_DNAVERSION)
+    )?;
+
+    Ok( (entry_hash.clone(), version.to_info( entry_hash )) )
+}
+
+
+
+
+#[derive(Debug, Deserialize)]
+pub struct GetDnaVersionInput {
+    pub addr: EntryHash,
+}
+
+#[hdk_extern]
+fn get_dna_version(input: GetDnaVersionInput) -> ExternResult<DnaVersionInfo> {
+    debug!("Get DNA Version: {}", input.addr );
+    let (_, element) = utils::fetch_entry_latest(input.addr.clone())?;
+
+    Ok(DnaVersionEntry::try_from(element)?.to_info( input.addr ))
+}
+
+
+
+
+fn get_version_links(dna: EntryHash) -> ExternResult<Vec<Link>> {
+    debug!("Getting version links for DNA: {}", dna );
+    let all_links: Vec<Link> = get_links(
+        dna,
+	Some(LinkTag::new(*TAG_DNAVERSION))
+    )?.into();
+
+    Ok(all_links)
+}
+
+
+#[derive(Debug, Deserialize)]
+pub struct GetDnaVersionsInput {
+    pub for_dna: EntryHash,
+}
+
+#[hdk_extern]
+fn get_dna_versions(input: GetDnaVersionsInput) -> ExternResult<Vec<(EntryHash, DnaVersionSummary)>> {
+    let links = get_version_links(input.for_dna)?;
+
+    let versions = links.into_iter()
+	.filter_map(|link| {
+	    match utils::fetch_entry_latest(link.target.clone()) {
+		Ok((_, element)) => Some((link.target, element)),
+		Err(_) => None
+	    }
+	})
+	.filter_map(|(hash, element)| {
+	    match DnaVersionEntry::try_from( element ) {
+		Err(_) => None,
+		Ok(version) => Some(( hash.clone(), version.to_summary(hash) )),
+	    }
+	})
+	.collect();
+    Ok( versions )
+}
+
+
+
+
+#[derive(Debug, Deserialize)]
+pub struct UpdateDnaVersionInput {
+    pub addr: EntryHash,
+    pub properties: DnaVersionUpdateOptions
+}
+#[derive(Debug, Deserialize)]
+pub struct DnaVersionUpdateOptions {
+    pub changelog: Option<String>,
+    pub contributors: Option<Vec<(String, Option<AgentPubKey>)>>,
+    pub published_at: Option<u64>,
+    pub last_updated: Option<u64>,
+}
+
+#[hdk_extern]
+fn update_dna_version(input: UpdateDnaVersionInput) -> ExternResult<(EntryHash, DnaVersionInfo)> {
+    debug!("Updating DNA Version: {}", input.addr );
+    let (header, element) = utils::fetch_entry_latest(input.addr.clone())?;
+    let current_version = DnaVersionEntry::try_from( element )?;
```

Comment:
Something that I have discovered is that if element contains sufficient bytes, and it unpacks into the target struct (a `DnaVersionEntry`, in this case) -- it will _successfully_ deserialize, even if it wasn't originally a DnaVersionEntry.  You actually need to do verification of the header content (eg. "is this a Header that could be a DnaVersionEntry").  You can only do this by examining the Header, unpacking the EntryType, and comparing the ZomeId and EntryDefIndex against the expected EntryType values:

```
    lazy_static! {
        static ref MUTUAL_ZID: ZomeId = zome_info().unwrap().zome_id;
        static ref PREA_EDI: EntryDefIndex = entry_def_index!(Preauth).unwrap();
        static ref XFER_EDI: EntryDefIndex = entry_def_index!(Transfer).unwrap();
    };
...
    if is_create_app_entry(&element, EntryVisibility::Public, *MUTUAL_ZID, *PREA_EDI) {
        if let Ok(prea) = Preauth::try_from(entry.to_owned()) {
...
}

/// Return an Element's EntryHash, iff it Creates the given EntryVisibility, ZomeId and EntryDefIndex
pub fn find_create_app_entry(
    element: &Element,
    visibility: EntryVisibility, // eg. ...::Public/Private
    zome_id: ZomeId,
    entry_id: EntryDefIndex
) -> Option<EntryHash> {
    if let Header::Create( header::Create{entry_hash, entry_type, ..}) = element.header() {
        if let EntryType::App(app_entry) = entry_type {
            if *app_entry.visibility() == visibility
                && app_entry.zome_id() == zome_id
                && app_entry.id() == entry_id
            {
                return Some(entry_hash.to_owned())
            }
        }
    };
    None
}

/// Affirm that an Element matches the given EntryVisibility, ZomeId and EntryDefIndex
pub fn is_create_app_entry(
    element: &Element,
    visibility: EntryVisibility, // eg. ...::Public/Private
    zome_id: ZomeId,
    entry_id: EntryDefIndex
) -> bool {
    find_create_app_entry( element, visibility, zome_id, entry_id ).is_some()
}
```